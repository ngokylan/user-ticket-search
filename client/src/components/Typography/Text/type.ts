import { ReactNode } from 'react';
import {
  TypographyDecoration,
  TypographyTransform,
  TypographyColor,
} from '../type';
export type TextHtmlTag = 'span' | 'strong' | 'small';
export type TextSize = 'lg' | 'rg' | 'sm' | 'xs' | 'button' | 'status';

export type TextProps = {
  /** ID of the component */
  id?: string;
  /** Children */
  children?: ReactNode;
  /** HTML tag to render the component as : 'span' | 'strong' | 'small' */
  htmlTag?: TextHtmlTag;
  /** Size that is defined in the Design System : 'lg' | 'rg' | 'sm' | 'xs' | 'button' | 'status' */
  size?: TextSize;
  /** It decorates its text node : 'overline', 'lineThrough', 'underline' */
  decoration?: TypographyDecoration;
  /** Style the text to be bold */
  isBold?: boolean;
  /** Style the text to be italic */
  isItalic?: boolean;
  /** Style the text node to transform into : 'lowercase', 'uppercase', 'capitalize' */
  transform?: TypographyTransform;
  /** Style the text node's colour into one of defined in Design System :
   * 'primary', 'secondary', 'black', 'gray', 'information', 'warning', 'success', 'danger'
   */
  color?: TypographyColor;
  /** class name attribute, (optional) */
  className?: string;
};
