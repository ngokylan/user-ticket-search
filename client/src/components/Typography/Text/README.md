Text wraps its nodes with one of inline html tag(span, strong and small) and provides styling.

HTML Tags
```js
<div><Text htmlTag="span">Text - Span </Text></div>
<div><Text htmlTag="strong">Text - Strong </Text></div>
<div><Text htmlTag="small">Text - Small </Text></div>
```

Sizes
```js
<div><Text size="lg">Text - Large</Text></div>
<div><Text size="rg">Text - Default</Text></div>
<div><Text size="sm">Text - Small</Text></div>
<div><Text size="xs">Text - Extra Small</Text></div>
<div><Text size="button">Text - Button</Text></div>
<div><Text size="status">Text - status</Text></div>
```

Please refer to <a href="#heading">Heading</a> for more examples.