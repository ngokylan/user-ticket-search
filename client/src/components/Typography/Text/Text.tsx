import React from 'react';
import StyledText from './style';
import { TextProps } from './type';
import { getClass } from '../../../_lib/helper';
import '../Typography.scss';

export function mapPropsToClass(block: string, props: TextProps) {
  const {
    isBold,
    isItalic,
    decoration,
    transform,
    color,
    size,
    className,
  } = props;
  return getClass(
    block,
    {
      [`${className}`]: className,
      [`font-weight-bold`]: isBold,
      [`font-italic`]: isItalic,
      [`text-decoration-${decoration}`]: decoration,
      [`text-transform-${transform}`]: transform,
      [`text-size-${size}`]: size,
      [`text-color-${color}`]: color,
    },
    []
  );
}

function Text({ children, ...props }: TextProps): any {
  const { id, htmlTag } = props;
  const StyledTextComponent: any = StyledText(htmlTag);

  return (
    <StyledTextComponent
      id={id}
      data-testid={`typography-${id || 'default'}`}
      className={mapPropsToClass('typography', props)}
    >
      {children}
    </StyledTextComponent>
  );
}

Text.defaultProps = {
  htmlTag: 'span',
  isBold: false,
  isItalic: false,
  isTruncate: false,
};

export default Text;
