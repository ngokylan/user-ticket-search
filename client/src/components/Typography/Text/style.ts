import styled from '../../../_lib/style';

const elements: any = {
  span: styled.span``,
  small: styled.small``,
  strong: styled.strong``,
};

const StyledText = function(htmlTag?: string) {
  return elements[htmlTag || 'span'];
};

export default StyledText;
