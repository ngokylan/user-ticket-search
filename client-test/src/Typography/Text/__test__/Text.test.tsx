import React from 'react';
import { mount } from 'enzyme';
import Text from '../';
import TextProps from '../';

const defaultProps: TextProps = {
  htmlTag: 'span',
  isBold: false,
  isItalic: false,
  isTruncate: false,
};

const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Text {...props} />);
  const component = wrapper.find('.typography');

  return { props, wrapper, component };
};

describe('Text', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Text.defaultProps)
        );
      });

      it.each([
        ['typography-default', {}],
        ['typography-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { wrapper } = mountComponent(props);
        expect(
          wrapper
            .find('.typography')
            .at(0)
            .prop('data-testid')
        ).toBe(expectedTestId);
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const text = 'Hello world';
        const { wrapper } = mountComponent({ children: text });
        expect(wrapper.find('span').text()).toBe(text);
      });
    });

    describe('custom', () => {
      let testOptions = [
        ['className', 'class-test', { className: 'class-test' }],
        ['isBold', 'font-weight-bold', { isBold: true }],
        ['isItalic', 'font-italic', { isItalic: true }],
      ];
      ['overline', 'lineThrough', 'underline'].map(function(value: string) {
        testOptions.push([
          'decoration',
          'text-decoration-' + value,
          { decoration: value },
        ]);
      });
      ['lowercase', 'uppercase', 'capitalize'].map(function(value: string) {
        testOptions.push([
          'transform',
          'text-transform-' + value,
          { transform: value },
        ]);
      });
      [
        'primary',
        'secondary',
        'black',
        'information',
        'warning',
        'success',
        'danger',
      ].map(function(value: string) {
        testOptions.push(['color', 'text-color-' + value, { color: value }]);
      });
      ['lg', 'rg', 'sm', 'xs', 'button', 'status'].map(function(value: string) {
        testOptions.push(['Text size', 'text-size-' + value, { size: value }]);
      });
      it.each(testOptions)(
        'should render class %s as %s',
        (name, expected, props) => {
          const { wrapper } = mountComponent(props);
          expect(wrapper.find('.' + expected).exists()).toEqual(true);
        }
      );
      it.each([
        ['span', { htmlTag: 'span' }],
        ['strong', { htmlTag: 'strong' }],
        ['small', { htmlTag: 'small' }],
      ])('should render with %s tag', (expected, props) => {
        const { wrapper } = mountComponent(props);
        const heading = wrapper.find(expected);
        expect(heading).toHaveLength(1);
      });
    });
  });
});
