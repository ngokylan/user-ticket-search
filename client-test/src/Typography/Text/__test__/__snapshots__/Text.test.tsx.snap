// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text rendering should be match with snapshot 1`] = `
<span
  className="sc-bdfBwQ typography"
  data-testid="typography-default"
/>
`;
